File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-non-null-asserted-optional-chain.test.ts
import { RuleTester, getFixturesRootDir } from '../RuleTester.ts';

const rootPath = getFixturesRootDir();

const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: ['./tsconfig.json'],
      tsconfigRootDir: rootPath,
    },
  },
});

ruleTester.run('no-non-null-asserted-optional-chain', {
  valid: [
    'foo.bar!;',
    'foo.bar!.baz;',
    'foo.bar!.baz();',
    'foo.bar()!;',
    'foo.bar()!();',
    'foo.bar()!.baz;',
    'foo?.bar;',
    'foo?.bar();',
    '(foo?.bar).baz!;',
    '(foo?.bar()).baz!;',
    'foo?.bar!.baz;',
    'foo?.bar!();',
    "foo?.['bar']!.baz;",
  ],
  invalid: [
    {
      code: 'foo?.bar!;',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: 'foo?.bar;',
            },
          ],
        },
      ],
    },
    {
      code: "foo?.['bar']!;",
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: "foo?.['bar'];",
            },
          ],
        },
      ],
    },
    {
      code: 'foo?.bar()!;',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: 'foo?.bar();',
            },
          ],
        },
      ],
    },
    {
      code: 'foo.bar?.()!;',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: 'foo.bar?.();',
            },
          ],
        },
      ],
    },
    {
      code: '(foo?.bar)!.baz',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: '(foo?.bar).baz',
            },
          ],
        },
      ],
    },
    {
      code: '(foo?.bar)!().baz',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: '(foo?.bar)().baz',
            },
          ],
        },
      ],
    },
    {
      code: '(foo?.bar)!',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: '(foo?.bar)',
            },
          ],
        },
      ],
    },
    {
      code: '(foo?.bar)!()',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: '(foo?.bar)()',
            },
          ],
        },
      ],
    },
    {
      code: '(foo?.bar!)',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: '(foo?.bar)',
            },
          ],
        },
      ],
    },
    {
      code: '(foo?.bar!)()',
      errors: [
        {
          messageId: 'noNonNullOptionalChain',
          suggestions: [
            {
              messageId: 'suggestRemovingNonNull',
              output: '(foo?.bar)()',
            },
          ],
        },
      ],
    },
  ],
});

File: internal/plugins/typescript/rules/no_non_null_asserted_optional_chain/no_non_null_asserted_optional_chain_test.go
package no_non_null_asserted_optional_chain

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoNonNullAssertedOptionalChainRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoNonNullAssertedOptionalChainRule, []rule_tester.ValidTestCase{
		{Code: `foo.bar!;`},
		{Code: `foo.bar!.baz;`},
		{Code: `foo.bar!.baz();`},
		{Code: `foo.bar()!;`},
		{Code: `foo.bar()!();`},
		{Code: `foo.bar()!.baz;`},
		{Code: `foo?.bar;`},
		{Code: `foo?.bar();`},
		{Code: `(foo?.bar).baz!;`},
		{Code: `(foo?.bar()).baz!;`},
		// Non-null assertion in middle of optional chain (not outermost)
		{Code: `foo?.bar!.baz;`},
		{Code: `foo?.bar!();`},
		{Code: `foo?.['bar']!.baz;`},
	}, []rule_tester.InvalidTestCase{
		// End of chain: foo?.bar!
		{
			Code: `foo?.bar!;`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `foo?.bar;`,
						},
					},
				},
			},
		},
		// End of chain: foo?.['bar']!
		{
			Code: `foo?.['bar']!;`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `foo?.['bar'];`,
						},
					},
				},
			},
		},
		// End of chain: foo?.bar()!
		{
			Code: `foo?.bar()!;`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `foo?.bar();`,
						},
					},
				},
			},
		},
		// End of chain: foo.bar?.()!
		{
			Code: `foo.bar?.()!;`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `foo.bar?.();`,
						},
					},
				},
			},
		},
		// Wrapping: (foo?.bar)!.baz
		{
			Code: `(foo?.bar)!.baz`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `(foo?.bar).baz`,
						},
					},
				},
			},
		},
		// Wrapping: (foo?.bar)!().baz
		{
			Code: `(foo?.bar)!().baz`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `(foo?.bar)().baz`,
						},
					},
				},
			},
		},
		// Wrapping: (foo?.bar)!
		{
			Code: `(foo?.bar)!`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `(foo?.bar)`,
						},
					},
				},
			},
		},
		// Wrapping: (foo?.bar)!()
		{
			Code: `(foo?.bar)!()`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    1,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `(foo?.bar)()`,
						},
					},
				},
			},
		},
		// Inside parens at end of chain: (foo?.bar!)
		{
			Code: `(foo?.bar!)`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    2,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `(foo?.bar)`,
						},
					},
				},
			},
		},
		// Inside parens at end of chain called: (foo?.bar!)()
		{
			Code: `(foo?.bar!)()`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noNonNullOptionalChain",
					Line:      1,
					Column:    2,
					Suggestions: []rule_tester.InvalidTestCaseSuggestion{
						{
							MessageId: "suggestRemovingNonNull",
							Output:    `(foo?.bar)()`,
						},
					},
				},
			},
		},
	})
}
